### PR TITLE
[DEV-36918] Suffix gargoyle cache key with py version

### PR DESCRIPTION
**Pull request by bal2ag**

### CODE REVIEW

**Review by Jaidan — DISMISSED** [review]

File: modeldict/base.py [hunk 1]
```diff
@@ -229,6 +230,16 @@ def _update_cache_data(self):
...
 
+        # Need to invalidate the cache key for the other Python version, so that
+        # it will get rewritten by a process running that Python version.
+        other_version_no = '2' if sys.version_info[0] == 3 else '3'
+        other_version_remote_cache_key = (
+            self.remote_cache_key[:-1]
+            + other_version_no
+        )
+
+        self.remote_cache.delete(other_version_remote_cache_key)
```
Comment (nh2d) [inline modeldict/base.py:252]: Can we write unit tests for this?

File: modeldict/models.py [hunk 2]
```diff
@@ -55,7 +57,9 @@ def __init__(self, model, key='pk', value=None, instances=False, auto_create=Fal
         self.instances = instances
         self.auto_create = auto_create
 
-        self.remote_cache_key = '%s:%s:%s:%s' % (cls_name, model_name, cache_version, self.key)
+        version_suffix = 'py%s' % sys.version_info[0]
+
+        self.remote_cache_key = '%s:%s:%s:%s:%s' % (cls_name, model_name, cache_version, self.key, version_suffix)
```
Comment (nh2d) [inline modeldict/models.py:62]: unit test?

**Review by nh2d — CHANGES_REQUESTED** [review]
LGTM. Couple comments about unit tests

**Comment by bal2ag** [comment]
@Jaidan @nh2d Let me summarize what this PR is trying to address.

In Python 3, if you try to unpickle an object that was pickled in Python 2 which contains `datetime`, `date`, or `time` objects inside it (e.g. as attributes), you get a `UnicodeDecodeError` unless you specify `encoding=latin1` when you call `pickle.load`.

Gargoyle stores the entire state of all switches (and associated condition sets) as a pickled object inside memcached. This object is unpickled inside `pylibmc` C code; I couldn't find an easy way to patch this code to pass the right encoding to handle those cases (somewhere here: https://github.com/lericson/pylibmc/blob/master/src/_pylibmcmodule.c#L2496).

So, the proposed solution is to have each Python version read and write its own copy of the Gargoyle memcached key. To make things work, whenever Gargoyle is updated (by a box  that could be running Python 2 or Python 3), both of these cache keys must be invalidated.

Gargoyle's normal flow on changes is to rewrite the `last_updated` value and the actual Gargoyle cache value at once, which ensures the Gargoyle cache key is present. If it's not present, the code falls back to the locally cached value until the process restarts.

To support version-suffixed keys, I added a change to treat a missing Gargoyle key as a cache invalidation by a box running the other Python version; in this case the code will refresh its version-suffixed key directly from the database.

I think this should work but it would be helpful if people can think about any failure modes they might expect to happen. This is definitely a bit of a scary change, but I think it's necessary to support simultaneous Python 2 and Python 3 compatibility.

File: modeldict/models.py (hunk 2 above)
Comment (bal2ag) [reply modeldict/models.py:62]: added

File: modeldict/base.py (hunk 1 above)
Comment (bal2ag) [reply modeldict/base.py:252]: added

**Comment by nh2d** [comment]
@bal2ag I'm not sure i fully understand the problem based on the description. 

So Gargoyle has a cache key to a value that represents a pickled object of the entire gargoyle state? e.g. `gargoyl_super_key -> picked_obj_with_all_state`

The solution thus is make sure to always read/write from the respective gargoyle key based on py version. E.g. py3 will use in memcached `gargoyle_super_key.py3 -> gargoyle_state_pickled` and py2 will use `gargoyle_super_key.py2 -> gargoyle_state_pickled`.

Making sure to read the right one seems trivial. But to update the value of anything within the pickled object (that represents state) you propose invalidating the cache for both versions on any machine that is attempting to update the pickled object. Normally the machine that attempts to update would pull the pickled object, make the change (? this part I don't understand, does it pull cache obj and update or does it read fresh from db?), and update `last_updated` and the cache_value in the same operation. This PR will change that flow to now invalidate both caches when a gargoyle change is detected. But which box(s) will be now responsible for updating the cache? The box for the next process that wants to check gargoyle?

Does it work like this: once both caches are invalidated any py2 box that tries to read from cache will miss then read from DB + update cache value, and any py3 box that tries to read will do the same? If this understanding is correct I have a few questions.

1. Can we trigger a read immediately after we invalidate to keep cache up-time higher for at least that py version?
2. I'm assuming it is safe for 2 attempted updates to the cache to occur at the same time. I.e. one box invalidates then 2+ other boxes try to read invalid cache and both simultaneously attempt updates. 
    a. This should be an existing failure mode when the state changes already so I'm not too worried
3. Is the same pool of hosts going to be reading from db and updating the cache that used to be doing it? I.e. are we introducing a db call (albeit small assuming it is just a read to get gargoyle state) to anywhere new that wasn't making these calls before?
    a. This is probably not a big deal assuming it is the same machines that previously would make the db call in the event of that gargoyle state changed and needed to update cache anyways.

**Review by nh2d — APPROVED** [review]
LGTM a few questions / comments that should be okay

**Comment by bal2ag** [comment]
@nh2d your understanding is mostly correct.

Current flow for writes (when Gargoyle changes):
1. Write changes to DB
2. Fetch entire state of Gargoyle from DB (`get_cache_data()`)
3. Write `remote_cache_key` (the "super key" whose value is entire Gargoyle state) and `remote_cache_last_updated` (which causes all other processes to update their local cache from the remote cache)

Current flow for reads:
1. Check local in-process cache
2. If expired (every 30 seconds +/- some jitter), compare `remote_cache_last_updated` with a local timestamp of when the last time the local cache was updated from the remote. If `remote_cache_last_updated` is greater, try to fetch entire state of Gargoyle from `remote_cache_key`. **If there is a cache miss, just continue falling back to local cache**. Otherwise rewrite local cache and mark last time updated from remote as now.
3. If local cache doesn't exist, treat it as expired.

What we are proposing changing is for each Python version to use it's own `remote_cache_key`. This necessitates two changes:
1. Because each `remote_cache_key` must be written by a process that is running the version of Python that will be reading from it, we must now delete the _other_ version's `remote_cache_key` when a process changes Gargoyle. This is necessary because the old flow just writes `remote_cache_key` directly, but now we have two of them.
2. If `remote_cache_key` is missing during the read flow, we assume that it was deleted from the "Gargoyle changed" flow on the other Python version and we rewrite it from the database (the next process in the Python version where the key is missing will do this).

To address your questions directly:
1) We can't trigger a read because the invalidation happens in the other Python version. E.g. Box running Python 2 changes the DB, writes the Python 2 `remote_cache_key`, and deletes the Python 3 `remote_cache_key`. Then the next Python 3 process to try to read it will find it missing and write it.
2) Yep, good question. Technically there is a narrow race condition if you make very rapid successive updates to the cache but I don't think this actually happens in practice.
3) Nothing changes here, except that it might be a Python 3 host or a Python 2 host.

**Pull request merged**